### PR TITLE
mgr/vol: handle case where clone index entry goes missing

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
In async_cloner.py, clone index entry is fetched to get next clone job
that needs to be executed. It might happen that the clone job was
cancelled just when it was going to be picked for execution (IOW, when
it was about to move from pending state to in-progress state).

Currently, in such a case MGR hangs because the exception ObjectNotFound
from CephFS Python bindings is raised and is left uncaught. To prevent
this issue catch an exception, log it and return None to tell get_job()
of async_job.py to look for next job in the queue.

Fixes: https://tracker.ceph.com/issues/66560

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
jenkins test make check arm64

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test make check arm64

File: src/pybind/mgr/volumes/fs/async_cloner.py [hunk 1]
```diff
@@ -37,6 +37,12 @@ def get_next_clone_entry(fs_client, volspec, volname, running_jobs):
                 if ve.errno == -errno.ENOENT:
                     return 0, None
                 raise ve
+            except cephfs.ObjectNotFound as e:
```
Comment (vshankar) [inline src/pybind/mgr/volumes/fs/async_cloner.py:40]: Isn't this case handle in `IndexException` above with the `ENOENT` check?

Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: It's not. If that was the case crash wouldn't have happened.

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: You didn't detail the crash (backtrace) nor a tracker was linked (until now!). How are reviewers supposed to know what is being fixed?

Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: The commit message described the issue and the fix in detail. The tracker wasn't linked because it wasn't tough to reproduce this and I had hit while I was testing something else. Please let me know if anything else is required.

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: Have you thought about expanding the try block to cover `lstat()` in `get_oldest_clone_entry()`?

**Review by vshankar — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/volumes/fs/async_cloner.py (hunk 1 above)
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: There's not lstat in async_cloner.py AFAIK...

Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: > But I think it'll be good idea to expand scope of that try-except so that future cases of cephfs.Error are caught.

Done.

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: Is this hunk really required given that get_oldest_clone_entry will throw IndexException when this case is seen?

Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: The quoted in this thread? Yes, it it required. `cephfs.ObjectNotFound` won't be caught unless we do.

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test make check arm64

**Comment by rishabh-d-dave** [comment]
jenkins test api

File: src/pybind/mgr/volumes/fs/async_cloner.py (hunk 1 above)
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: @vshankar ping

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: Why? `clone_index.get_oldest_clone_entry()` throws IndexException and the ENOENT check would be covered.

Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py:40]: `get_oldest_clone_entry()` throws `IndexException` for `cephfs.Error` but `ObjectNotFound` is not inherited from `cephfs.Error`. So, `ObjectNotFound` would be left uncaught and would rise up callstack until caught.

File: src/pybind/mgr/volumes/fs/operations/clone_index.py [hunk 2]
```diff
@@ -66,6 +66,8 @@ def get_oldest_clone_entry(self, exclude=[]):
             return None
         except cephfs.Error as e:
             raise IndexException(-e.args[0], e.args[1])
+        except cephfs.ObjectNotFound as e:
+            raise IndexException(-e.args[0], e.args[1])
```
Comment (rishabh-d-dave) [inline src/pybind/mgr/volumes/fs/operations/clone_index.py:70]: @vshankar 
Because -

```
class ObjectNotFound(OSError):
    pass
```

File: src/pybind/mgr/volumes/fs/async_cloner.py (hunk 1 above)
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py]: See https://github.com/ceph/ceph/pull/58101#discussion_r1648074192

File: src/pybind/mgr/volumes/fs/async_cloner.py [hunk 3]
```diff
@@ -33,10 +33,10 @@ def get_next_clone_entry(fs_client, volspec, volname, running_jobs):
                 with open_clone_index(fs_handle, volspec) as clone_index:
                     job = clone_index.get_oldest_clone_entry(running_jobs)
                     return 0, job
-            except IndexException as ve:
-                if ve.errno == -errno.ENOENT:
-                    return 0, None
-                raise ve
+        except IndexException as ve:
```
Comment (vshankar) [inline src/pybind/mgr/volumes/fs/async_cloner.py:36]: unrelated change, please revert.

**Review by vshankar — CHANGES_REQUESTED** [review]
otherwise lgtm.

File: src/pybind/mgr/volumes/fs/async_cloner.py (hunk 3 above)
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py]: Done.

File: src/pybind/mgr/volumes/fs/async_cloner.py (hunk 1 above)
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/async_cloner.py]: @vshankar Can I mark this thread as resolved?

**Comment by rishabh-d-dave** [comment]
@vshankar I've squashed the commits and made the change you had suggested. It's ready for review. PTAL.

File: src/pybind/mgr/volumes/fs/operations/clone_index.py [hunk 4]
```diff
@@ -64,6 +64,16 @@ def get_oldest_clone_entry(self, exclude=[]):
                 sink_path = self.fs.readlink(os.path.join(self.path, min_ctime_entry[0]), CloneIndex.PATH_MAX)
                 return (min_ctime_entry[0], sink_path[:linklen])
             except cephfs.Error as e:
+                log.debug('Exception cephfs.Error has been caught. It appears that '
```
Comment (vshankar) [inline src/pybind/mgr/volumes/fs/operations/clone_index.py]: The debug logs are same for this and `cephfs.ObjectNotFound` which does not make sense.

Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: Fixed.

Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: ping

**Comment by rishabh-d-dave** [comment]
@vshankar Just a reminder, please re-review.

**Review by vshankar — APPROVED** [review]

**Comment by rishabh-d-dave** [comment]
jenkins test make check

File: src/pybind/mgr/volumes/fs/operations/clone_index.py [hunk 5]
```diff
@@ -64,6 +64,14 @@ def get_oldest_clone_entry(self, exclude=[]):
                 sink_path = self.fs.readlink(os.path.join(self.path, min_ctime_entry[0]), CloneIndex.PATH_MAX)
                 return (min_ctime_entry[0], sink_path[:linklen])
             except cephfs.Error as e:
```
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: Ignore the above comment, I see that's been handled in the second commit 9ef9e17dfcbf6df76823f812441992bb5e87ed93 
Isn't it a better idea to squash both commit into one ? What's the advantage of splitting this into two especially when the same author is involved? Won't the commit history of the file become clumsy ?

[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: I don't mind squashing, I kept them as two separate commits because  first one fixes the bug and second is actually an improvement that increases scope of try-except block.

[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: I suggest to squash them into single commit.

[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: Okay, in that case it will be hard to point in future that exactly what fixed the bug -- extending scope try-except or catching ObjectNotFound or combination of both. Right now it's clear that only first one was sufficient to catch it.

[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: Now I see the actual traceback in the tracker.

```
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_job.py", line 52, in run vol_job = self.async_job.get_job()
^^^^^^^^^^^^^^^^^^^^^^^^
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_job.py", line 195, in get_job (ret, job) = self.get_next_job(volname, running_jobs)
^^^^^^^^^^^^^^^.
^^^^^^^^^^^^^^^
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_cloner.py", line 402, in get_next_job return get_next_clone_entry(self.fs_client, self.vc.volspec, volname, running_jobs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/async_cloner.py", line 35, in get_next_clone_entry job = clone_index.get_oldest_clone_entry(running_jobs)
ΑΛΛΛΛΛΛΛ
ΜΑΛΛΙ
ile "/home/rishabh/repos/ceph/mgr-vol-clone-stats/src/pybind/mgr/volumes/fs/operations/clone_index.py", line 80, in get_oldest_clone_en
st = self.fs.lstat(dpath)
ΑΛΛΛΛΛΛ
ile "cephfs.pyx", line 1997, in cephfs.LibCephFS.lstat
ile "cephfs.pyx", line 1978, in cephfs.LibCephFS.stat
hfs.ObjectNotFound: error in stat: /volumes/_index/clone/3efa5bf7-115c-4f27-8b83-8fecc16a690a: No such file or directory [Errno 2]
```

It is very clearly logged that the `ile "cephfs.pyx", line 1978, in cephfs.LibCephFS.stat` called from ` get_oldest_clone_entry
st = self.fs.lstat(dpath) `is raising the exception cephfs.ObjectNotFound.

The only instance of lstat is at line no. 57 (https://github.com/ceph/ceph/commit/60821ecfd32cb9332713108e1ad6ec3a1ae8a8c9#diff-32b8a30d8ea6ea9fd5b4753c03bfa92c14dd72aab44cdb973856448514d0b320R57)

I think the whole confusion started because of `cephfs.ObjectNotFound` in the traceback. The general coding standard/approach followed in mgr/volumes is to catch `cephfs.Error` when a call is made to libcephfs. Please note that `cephfs.Error' is the parent exception class of all the exceptions raised by python binding call to libcephfs (cephfs.pyx). So `cephfs.Error` would catch any exception from libcephfs and the code relies on `ret` value which is actual linux errno to decide on how to treat the exception.

So there is only one bug here, i.e., the exception from `self.fs.lstat` is not being caught. 
The `self.fs.readlink` exception is being caught correctly and the code added to catch `cephfs.ObjectNotFound` will be dead code, since `cephfs.Error` would catch it before.

File: src/pybind/mgr/volumes/fs/operations/clone_index.py [hunk 6]
```diff
@@ -48,32 +48,32 @@ def untrack(self, tracking_id):
...
+        except cephfs.Error as e:
+            log.debug('Exception cephfs.Error has been caught. Printing '
+                      f'the exception - {e}')
+            raise IndexException(-e.args[0], e.args[1])
+        except cephfs.ObjectNotFound as e:
+            log.debug('Exception cephfs.ObjectNotFound has been caught. It '
+                      'appears that a clone index entry was removed perhaps '
+                      'because that clone was cancelled by the user. Printing '
+                      f'the exception - {e}')
+            raise IndexException(-e.args[0], e.args[1])
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/operations/clone_index.py:76]: This is dead code. Please remove.

File: src/pybind/mgr/volumes/fs/operations/clone_index.py [hunk 7]
```diff
@@ -48,32 +48,32 @@ def untrack(self, tracking_id):
...
-                log.debug('Exception cephfs.ObjectNotFound has been caught. It '
-                          'appears that a clone index entry was removed perhaps '
-                          'because that clone was cancelled by the user. Printing '
-                          f'the exception - {e}')
-                raise IndexException(-e.args[0], e.args[1])
-        return None
+            return None
+        except cephfs.Error as e:
+            log.debug('Exception cephfs.Error has been caught. Printing '
+                      f'the exception - {e}')
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/operations/clone_index.py:69]: The caller is handling the case of ENOENT and we should be good. Not sure if the debug message is required here.

File: src/pybind/mgr/volumes/fs/operations/clone_index.py [hunk 8]
```diff
@@ -48,32 +48,32 @@ def untrack(self, tracking_id):
...
-        if min_ctime_entry:
-            try:
+        try:
+            min_ctime_entry = None
+            exclude_tracking_ids = [v[0] for v in exclude]
+            log.debug("excluded tracking ids: {0}".format(exclude_tracking_ids))
+            for entry in list_one_entry_at_a_time(self.fs, self.path):
+                dname = entry.d_name
+                dpath = os.path.join(self.path, dname)
+                st = self.fs.lstat(dpath)
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/operations/clone_index.py:58]: All we require is to just include this line under `try` block retaining the old `except` block.

**Review by kotreshhr — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/volumes/fs/operations/clone_index.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: > . Please note that cephfs.Error is the parent exception class of all the exceptions(including cephfs.ObjectNotFound) 

Copying from `src/pybind/cephfs/cephfs.pyx` -

```
class Error(Exception):
    def get_error_code(self):
        return 1
```

```
class ObjectNotFound(OSError):
    pass
```

File: src/pybind/mgr/volumes/fs/operations/clone_index.py (hunk 7 above)
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py:69]: This helps finding that an exception was raised while reading the logs.

File: src/pybind/mgr/volumes/fs/operations/clone_index.py (hunk 6 above)
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py:76]: See: https://github.com/ceph/ceph/pull/58101#discussion_r1657753917

File: src/pybind/mgr/volumes/fs/operations/clone_index.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: Yes and `OSError` has inherited `Error` class

```
class OSError(Error): 
    def __init__(self, errno, strerror):
        super(OSError, self).__init__(errno, strerror)
        self.errno = errno
        self.strerror = "%s: %s" % (strerror, os.strerror(errno))
    
    def __str__(self):
        return '{} [Errno {}]'.format(self.strerror, self.errno)
    
    def get_error_code(self):
        return self.errno

```

File: src/pybind/mgr/volumes/fs/operations/clone_index.py (hunk 6 above)
Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: See: https://github.com/ceph/ceph/pull/58101#discussion_r1658129186

**Review by kotreshhr — APPROVED** [review]

File: src/pybind/mgr/volumes/fs/operations/clone_index.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply src/pybind/mgr/volumes/fs/operations/clone_index.py]: Yes. I've made the change yesterday itself. :)

**Review by rishabh-d-dave — COMMENTED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#11-JUL-2024

Testing took some time because first there were issues with building testing branch, then there were too many infra failures and finally last run's rerun had new failures that need to be checked. There were no failures related to this PR, approving this PR.

**Comment by rishabh-d-dave** [comment]
jenkins test make check

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
I am seeing make check failures that seem unrelated - https://jenkins.ceph.com/job/ceph-pull-requests/138712/consoleFull#-1640523647e840cee4-f4a4-4183-81dd-42855615f2c1. Copying them below -

```
Identified `/home/jenkins-build/build/workspace/ceph-pull-requests` as project root containing a .git directory.
Found input source directory: "/home/jenkins-build/build/workspace/ceph-pull-requests/src/cephadm/cephadmlib"

Aborted!
check-black: exit 1 (0.45 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/src/cephadm> black --check -l78 -t py36 --skip-string-normalization -v cephadmlib/ pid=3832823
check-black: FAIL ✖ in 11.96 seconds
```

```
check-black: commands[0] /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr> black --check -q -l78 -t py36 --skip-string-normalization smb

Aborted!
check-black: exit 1 (0.27 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr> black --check -q -l78 -t py36 --skip-string-normalization smb pid=3951087
check-black: FAIL ✖ in 4.17 seconds
```

**Comment by rishabh-d-dave** [comment]
make check passed by rebasing PR branch.

**Comment by rishabh-d-dave** [comment]
jenkins test api

**Comment by rishabh-d-dave** [comment]
CI passed, proceeding to merge. Leaving link to the comment where QA run reported - https://github.com/ceph/ceph/pull/58101#pullrequestreview-2171517047
